pynumaflow/batchmapper/async_server.py

```diff
@@ -0,0 +1,213 @@
+            ("grpc.max_receive_message_length", self.max_message_size),
+        ]
+
+        self.servicer = servicer
+
+    def start(self) -> None:
+        """
+        Starter function for the Async server class, need a separate caller
+        so that all the async coroutines can be started from a single context
+        """
+        print("start")
+        aiorun.run(self.aexec(), use_uvloop=True)
+
+    async def aexec(self) -> None:
+        """
+        Starts the Async gRPC server on the given UNIX socket with
+        given max threads.
+        """
+
+        # As the server is async, we need to create a new server instance in the
+        # same thread as the event loop so that all the async calls are made in the
+        # same context
+        server_new = grpc.aio.server()
+        server_new.add_insecure_port(self.sock_path)
+        batchmap_pb2_grpc.add_BatchMapServicer_to_server(self.servicer, server_new)
+
+        # Start the async server
+        await start_async_server(
+            server_new,
+            self.sock_path,
+            self.max_threads,
+            self._server_options,
+            self.server_info_file,
+        )
+
+
+class BatchMapServer(BatchMapAsyncServerBase):
+    """
+    This creates a Server instance to handler an input stream of data and
+    allow data to be streamed back out. The input is not bounded at the SDK level
+    and handler function must be cognizant to handle to handle each message and to decide
+    access patterns.
+
+    Example invocation:
+        from pynumaflow.batchmapper import (
+            Message,
+            Messages,
+            Datum,
+            BatchMapper,
+            BatchMapServer,
+            BatchResponses,
+        )
+
+        class MapperStreamer(BatchMapper):
+            async def handler(self, datum: AsyncIterable[Datum]) -> AsyncIterable[BatchResponses]:
+                async for msg in datum:
+                    msgs = Messages(
+                        Message(value=msv.value, keys=[], tags=[])
+                    )
+                    response = BatchResponses(msg.id, msgs)
+                    yield response
+
+        if __name__ == "__main__":
+            handler = MapperStreamer()
+            grpc_server = BatchMapServer(handler)
+            grpc_server.start()
+    """
+
+    def __init__(
+        self,
+        mapper_instance: MapBatchAsyncCallable,
+        sock_path=MAP_BATCH_SOCK_PATH,
+        max_message_size=MAX_MESSAGE_SIZE,
+        max_threads=MAX_THREADS,
+        server_info_file=MAP_BATCH_SERVER_INFO_FILE_PATH,
+    ):
+        servicer = BatchMapServicer(mapper_instance)
+        super().__init__(
+            servicer,
+            sock_path=sock_path,
+            max_message_size=max_message_size,
+            max_threads=max_threads,
+            server_info_file=server_info_file,
+        )
+
+
+class BatchMapUnaryServer(BatchMapAsyncServerBase):
```

Comment:
Let's keep batchmap to only use the batch interface (N messages together).
While I am in agreement with such a functionality which mimics unary (and would be a target state eventually), we could keep this part of a separate change where we can change the underlying system for map to bi-di completely. This will be in the pipeline as well soon